merkle-patricia-tree/src/Data/Tree/MerklePatricia.hs

```diff
@@ -98,7 +103,51 @@ instance HashAlgorithm alg => Eq (MerklePatriciaTree alg a) where
 -- | An opaque 'Proof' which can be used to prove membership or atomic
 -- operations such as 'add' or 'delete'
 newtype Proof alg = Proof {unProof :: [([Bit], [(Bit, Digest alg)])]}
-  deriving stock (Show)
+  deriving stock (Eq, Show)
+
+instance HashAlgorithm alg => Serialise (Proof alg) where
+  encode (Proof prf) = encodeListLen (fromIntegral $ length prf) <> mconcat (map encodeItem prf)
+   where
+    encodeItem (bits, digests) =
+      -- NOTE(AB): we need to pad the bits length to 8-bit words multiples, but we want
+      -- to keep track of the original length for deserialisation
+      -- TODO: Find a better encoding for the Proof object itself
+      encodeListLen (fromIntegral $ length bits)
+        <> encodeBytes (unsafeFromBits $ pad bits)
```

Comment:
It would actually be _as compact_ to encode them as a list of uint8 right? And, no need to pad them this way or am I missing something :thinking: ?

Reply:
My sentence might not be clear: I was referring to how they are "encoded" in the `Proof` type, not how they are serialised. So yes, if the bits were already stored as a `ByteString` or `[Word8]` then there would be no need for padding. Or did I misunderstood what you said?

Reply:
Aaaaaaah. Yes. That's my bad, the `unsafeFromBits` does indeed compact the bits into a sequence of `Word8`, by groups of 8. However, there shouldn't be any need for padding right? ... provided that the proof was constructed via the legitimate ways of course! Perhaps you did hit this corner case when generating arbitrary proofs :thinking:  ?

Reply:
I don't think so: The proofs are generated from an arbitrary MPT but produced through `mkProof`

Reply:
I am using `forAllNonEmptyMPT` which uses `unsafeMkProof` which uses `mkProof`